File: 백준/Silver/14501. 퇴사/퇴사.py
n = int(input()) # 일 할 수 있는 날짜 입력받기
job = []

for i in range(n) :
  t, p = map(int, input().split()) # 상담 기간과 금액 입력받기
  
  job.append((t, p)) # job에 튜플 추가

dp = [0] * (n + 1)

for i in range(n - 1, -1, -1) : # job을 역순으로 확인
  t, p = job[i]
  
  if i + t <= n : # 만약 이 일을 완료할 수 있다면
    dp[i] = max(p + dp[i + t], dp[i + 1]) # 더 큰 이익을 저장
  else :
    dp[i] = dp[i + 1] # 그대로 이전 값을 저장
    
print(dp[0]) # 최종 이익 출력

File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        List<int[]> list = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            list.add(Arrays.stream(br.readLine().split(" "))
                    .mapToInt(Integer::parseInt)
                    .toArray());
        }

        int[] earned = new int[n + 1];
        
        // 역순으로 최대 수익 계산
        for (int i = n - 1; i >= 0; i--) {
            int duration = list.get(i)[0];
            int profit = list.get(i)[1];

            // 일을 선택할 수 없는 경우
            if (i + duration > n) {
                earned[i] = earned[i + 1];
            } else {
                // 일을 선택했을 때의 수익과 일을 선택하지 않았을 때의 수익 중 더 큰 값
                earned[i] = Math.max(profit + earned[i + duration], earned[i + 1]);
            }
        }

        System.out.println(earned[0]);
    }
}
